internal/events/subscription.go

```diff
@@ -168,12 +173,39 @@ func (s *subscription) restartFilter(ctx context.Context, since *big.Int) error
 	if err != nil {
 		return errors.Errorf(errors.RPCCallReturnedError, "eth_newFilter", err)
 	}
+	s.catchupBlock = nil // we are not in catchup mode now
 	s.filteredOnce = false
 	s.markFilterStale(ctx, false)
 	log.Infof("%s: created filter from block %s: %s - %+v", s.logName, since.String(), s.filterID.String(), s.info.Filter)
 	return err
 }
 
+func (s *subscription) restartFilter(ctx context.Context, checkpoint *big.Int) error {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	since := checkpoint
+	if s.catchupBlock != nil {
+		// If we're already in catchup mode, we need to look at the current catchupBlock,
+		// not the checkpoint.
+		since = s.catchupBlock
+	}
+
+	blockNumber := ethbinding.HexBigInt{}
+	err := s.rpc.CallContext(ctx, &blockNumber, "eth_blockNumber")
+	if err != nil {
+		return errors.Errorf(errors.RPCCallReturnedError, "eth_blockNumber", err)
+	}
+
+	blockGap := new(big.Int).Sub(blockNumber.ToInt(), since).Int64()
+	if s.catchupModeBlockGap > 0 && blockGap > s.catchupModeBlockGap {
```

Comment:
Just a note that this will be a no-op and we won't try to re-catch up when a node is being re-synced or if there's an accidental chain rewind to a historical block with state after a non-graceful shutdown of the node. Think that is the right behavior given there's a stored checkpoint that tells us those blocks are already processed, maybe worth adding a comment.